src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs

```diff
@@ -44,17 +44,19 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
                 return new BoundNullCoalescingOperator(syntax, rewrittenLeft, rewrittenRight, rewrittenConversion, rewrittenResultType);
             }
 
+            var isUnconstrainedTypeParameter = rewrittenLeft.Type != null && !rewrittenLeft.Type.IsReferenceType && !rewrittenLeft.Type.IsValueType;
+
             // first we can make a small optimization:
             // If left is a constant then we already know whether it is null or not. If it is null then we 
             // can simply generate "right". If it is not null then we can simply generate
-            // MakeConversion(left).
-
-            if (rewrittenLeft.IsDefaultValue())
+            // MakeConversion(left). This does not hold when the left is an unconstrained type parameter: at runtime,
+            // it can be either left or right depending on the runtime type of T
+            if (rewrittenLeft.IsDefaultValue() && !isUnconstrainedTypeParameter)
```

Comment:
>!isUnconstrainedTypeParameter [](start = 50, length = 29)

It feels like it will be cheaper to check ```!isUnconstrainedTypeParameter``` first, and perhaps only once for the purpose of this ```if``` and the next one. Something like:
```
if (!isUnconstrainedTypeParameter)
{
            if (rewrittenLeft.IsDefaultValue())
            {
                return rewrittenRight;
            }

            if (rewrittenLeft.ConstantValue != null)
            {
                Debug.Assert(!rewrittenLeft.ConstantValue.IsNull);

                return GetConvertedLeftForNullCoalescingOperator(rewrittenLeft, leftConversion, rewrittenResultType);
            }
}
```